controllers/kubeadmconfig_controller_test.go

```diff
@@ -554,6 +561,180 @@ func TestReconcileIfJoinNodesAndControlPlaneIsReady(t *testing.T) {
 	if cfg.Status.BootstrapData == nil {
 		t.Fatal("Expected status ready")
 	}
+
+	l, err := myremoteclient.List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(fmt.Sprintf("Failed to reconcile:\n %+v", err))
+	}
+
+	if len(l.Items) != 2 {
+		t.Fatal(fmt.Sprintf("Failed to reconcile:\n %+v", err))
+	}
+}
+
+func TestReconcileDiscovery(t *testing.T) {
+	myremoteclient := fakeclient.NewSimpleClientset().CoreV1().Secrets(metav1.NamespaceSystem)
+
+	k := &KubeadmConfigReconciler{
+		Log:    log.ZapLogger(true),
+		Client: nil,
+		secretInterfaceFactory: func(cluster *capiv1alpha2.Cluster) (corev1.SecretInterface, error) {
+			return myremoteclient, nil
+		},
+	}
+
+	dummyCAHash := []string{"...."}
+	goodcluster := &capiv1alpha2.Cluster{
+		Status: capiv1alpha2.ClusterStatus{
+			APIEndpoints: []capiv1alpha2.APIEndpoint{
+				{
+					Host: "foo.com",
+					Port: 6443,
+				},
+			},
+		},
+	}
+	var useCases = []struct {
+		name              string
+		cluster           *capiv1alpha2.Cluster
+		config            *cabpkV1alpha2.KubeadmConfig
+		validateDiscovery func(*testing.T, *cabpkV1alpha2.KubeadmConfig)
+		expectedError     bool
```

Comment:
I'm not picky on tests so this is non-blocking, but I wanted to offer some advice:

Make unit tests test exactly one thing. I have found the complexity of tests becomes unwieldy if this advice is not followed.

This one function could be at *least* two different `func Test..()` functions.

One of them could be `TestReconcileDiscoverSuccess` and `TestReconcilceDiscoverErrors`. This way your test logic has *no* conditionals.